tools/kth_threshold.cpp

```diff
@@ -0,0 +1,211 @@
+#include <iostream>
+#include <optional>
+#include <unordered_set>
+
+#include "boost/algorithm/string/classification.hpp"
+#include "boost/algorithm/string/split.hpp"
+#include <boost/functional/hash.hpp>
+
+#include "mio/mmap.hpp"
+#include "spdlog/sinks/stdout_color_sinks.h"
+#include "spdlog/spdlog.h"
+
+#include "mappable/mapper.hpp"
+
+#include "app.hpp"
+#include "cursor/max_scored_cursor.hpp"
+#include "index_types.hpp"
+#include "io.hpp"
+#include "query/algorithm.hpp"
+#include "util/util.hpp"
+#include "wand_data_compressed.hpp"
+#include "wand_data_raw.hpp"
+
+#include "query/algorithm.hpp"
+#include "scorer/scorer.hpp"
+
+#include "CLI/CLI.hpp"
+
+using namespace pisa;
+
+template <typename IndexType, typename WandType>
+void kt_thresholds(
+    const std::string& index_filename,
+    const std::string& wand_data_filename,
+    const std::vector<Query>& queries,
+    std::string const& type,
+    ScorerParams const& scorer_params,
+    uint64_t k,
+    bool quantized,
+    std::optional<std::string> pairs_filename,
+    std::optional<std::string> triples_filename)
+{
+    IndexType index;
+    mio::mmap_source m(index_filename.c_str());
+    mapper::map(index, m);
+
+    WandType wdata;
+
+    auto scorer = scorer::from_params(scorer_params, wdata);
+
+    mio::mmap_source md;
+    std::error_code error;
+    md.map(wand_data_filename, error);
+    if (error) {
+        spdlog::error("error mapping file: {}, exiting...", error.message());
+        std::abort();
+    }
+    mapper::map(wdata, md, mapper::map_flags::warmup);
+
+    using Pair = std::set<uint32_t>;
+    std::unordered_set<Pair, boost::hash<Pair>> pairs_set;
+
+    using Triple = std::set<uint32_t>;
+    std::unordered_set<Triple, boost::hash<Triple>> triples_set;
+
+    std::string t;
+    if (pairs_filename) {
+        std::ifstream pin(*pairs_filename);
+        while (std::getline(pin, t)) {
+            std::vector<std::string> p;
+            boost::algorithm::split(p, t, boost::is_any_of(" \t"));
+            pairs_set.insert({(uint32_t)std::stoi(p[0]), (uint32_t)std::stoi(p[1])});
```

Comment:
This will cause headaches sooner or later. I suggest the following:
- Extract this logic to a function and parameterize it with the number of terms expected (such that we can use it below for triples).
- Introduce one exception class that will report invalid format (could be just `std::runtime_error` with appropriate messages).
- Throw if the size of the vector isn't what we expect.
- Catch both `std::invalid_argument` and `std::out_of_range` that `std::stoi` throws, and re-throw same type as above with a good message.
- Make sure in both cases to include what went wrong (type of error) and the line at which it failed, so that a user can find what went wrong easily.

Reply:
I have changed it. Please check this is what you mean